### PR TITLE
Remove pattern-based middleware

### LINKED ISSUE #1566: Remove pattern-based middleware
@sungam3r  We discussed removing pattern-based middleware previously, and the thought was that we were trying to maintain compatibility with older third-party middleware libraries.

I've been working on the upgrade guide, and realized that middleware designed for 2.x will not work with 3.0 due to the signature of the Resolve method having changed (`ResolveFieldContext` changed to `IResolveFieldContext`).

I suggest that we either:
- Remove the pattern-based middleware approach entirely, or
- Change the pattern-based middleware approach to scan only for a `Resolve` method with a `ResolveFieldContext` parameter - which exists for backwards-compatibility purposes only

Either way, I believe we should update the documentation to only document the current recommended technique using `IFieldMiddleware`.  The upgrade guide can cover backwards compatibility, if we choose to support it.

Please let me know what you think.

### PR DESCRIPTION
**Pull request by Shane32**
fixes #1566

### CODE REVIEW

File: src/GraphQL/Instrumentation/FieldMiddlewareBuilderExtensions.cs [hunk 1]
```diff
@@ -12,10 +12,8 @@ namespace GraphQL.Instrumentation
     /// </summary>
     public static class FieldMiddlewareBuilderExtensions
     {
-        private const string INVOKE_METHOD_NAME = "Resolve";
-
         /// <summary>
-        /// Adds middleware to the list of delegates that will be applied to the schema when invoking <see cref="ApplyTo(ISchema)"/>.
+        /// Adds middleware to the list of delegates that will be applied to the schema when invoking <see cref="FieldMiddlewareBuilder.ApplyTo(ISchema)"/>.
```
Comment (sungam3r) [inline src/GraphQL/Instrumentation/FieldMiddlewareBuilderExtensions.cs]:
```suggestion
        /// Adds middleware to the list of delegates that will be applied to the schema when invoking <see cref="IFieldMiddlewareBuilder.ApplyTo(ISchema)"/>.
```

Comment (sungam3r) [reply src/GraphQL/Instrumentation/FieldMiddlewareBuilderExtensions.cs]: and other places too

File: src/GraphQL/Instrumentation/FieldMiddlewareBuilderExtensions.cs [hunk 2]
```diff
@@ -53,72 +51,24 @@ public static IFieldMiddlewareBuilder Use(this IFieldMiddlewareBuilder builder,
...
-            static Exception NotSupported(ISchema schema) => new NotSupportedException($"'{schema.GetType().FullName}' should implement 'IServiceProvider' interface for resolving middlewares.");
-
-            static T CheckNotNull<T>(T instance, System.Type type)
-            {
-                if (instance == null)
-                    throw new InvalidOperationException($"Field middleware of type '{type.FullName}' must be registered in DI container.");
-                return instance;
-            }
+            if (!typeof(IFieldMiddleware).IsAssignableFrom(middleware))
+                throw new ArgumentException($"Field middleware of type '{middleware.FullName}' must implement {nameof(IFieldMiddleware)}", nameof(middleware));
```
Comment (sungam3r) [inline src/GraphQL/Instrumentation/FieldMiddlewareBuilderExtensions.cs]:
```suggestion
                throw new ArgumentException($"Field middleware of type '{middleware.FullName}' must implement {nameof(IFieldMiddleware)} interface", nameof(middleware));
```

File: src/GraphQL/Instrumentation/FieldMiddlewareBuilderExtensions.cs [hunk 3]
```diff
@@ -24,7 +22,7 @@ public static IFieldMiddlewareBuilder Use(this IFieldMiddlewareBuilder builder,
             => builder.Use(next => context => middleware.Resolve(context, next));
 
         /// <summary>
-        /// Adds the specified delegate to the list of delegates that will be applied to the schema when invoking <see cref="ApplyTo(ISchema)"/>.
+        /// Adds the specified delegate to the list of delegates that will be applied to the schema when invoking <see cref="FieldMiddlewareBuilder.ApplyTo(ISchema)"/>.
```
Comment (Shane32) [inline src/GraphQL/Instrumentation/FieldMiddlewareBuilderExtensions.cs]:
```suggestion
        /// Adds the specified delegate to the list of delegates that will be applied to the schema when invoking <see cref="IFieldMiddlewareBuilder.ApplyTo(ISchema)"/>.
```

File: src/GraphQL/Instrumentation/FieldMiddlewareBuilderExtensions.cs [hunk 4]
```diff
@@ -34,17 +32,17 @@ public static IFieldMiddlewareBuilder Use(this IFieldMiddlewareBuilder builder,
             => builder.Use((_, next) => middleware(next));
 
         /// <summary>
-        /// Adds middleware specified by its type to the list of delegates that will be applied to the schema when invoking <see cref="ApplyTo(ISchema)"/>.
+        /// Adds middleware specified by its type to the list of delegates that will be applied to the schema when invoking <see cref="FieldMiddlewareBuilder.ApplyTo(ISchema)"/>.
```
Comment (Shane32) [inline src/GraphQL/Instrumentation/FieldMiddlewareBuilderExtensions.cs]:
```suggestion
        /// Adds middleware specified by its type to the list of delegates that will be applied to the schema when invoking <see cref="IFieldMiddlewareBuilder.ApplyTo(ISchema)"/>.
```

File: src/GraphQL/Instrumentation/FieldMiddlewareBuilderExtensions.cs [hunk 5]
```diff
@@ -34,17 +32,17 @@ public static IFieldMiddlewareBuilder Use(this IFieldMiddlewareBuilder builder,
...
         /// </summary>
         /// <typeparam name="T">Middleware type.</typeparam>
         /// <param name="builder">Interface for connecting middlewares to a schema.</param>
         /// <returns>Reference to the same <see cref="IFieldMiddlewareBuilder"/>.</returns>
-        public static IFieldMiddlewareBuilder Use<T>(this IFieldMiddlewareBuilder builder) => Use(builder, typeof(T));
+        public static IFieldMiddlewareBuilder Use<T>(this IFieldMiddlewareBuilder builder) where T : IFieldMiddleware => Use(builder, typeof(T));
 
         /// <summary>
-        /// Adds middleware specified by its type to the list of delegates that will be applied to the schema when invoking <see cref="ApplyTo(ISchema)"/>.
+        /// Adds middleware specified by its type to the list of delegates that will be applied to the schema when invoking <see cref="FieldMiddlewareBuilder.ApplyTo(ISchema)"/>.
```
Comment (Shane32) [inline src/GraphQL/Instrumentation/FieldMiddlewareBuilderExtensions.cs]:
```suggestion
        /// Adds middleware specified by its type to the list of delegates that will be applied to the schema when invoking <see cref="IFieldMiddlewareBuilder.ApplyTo(ISchema)"/>.
```

File: src/GraphQL/Instrumentation/FieldMiddlewareBuilderExtensions.cs (hunk 2 above)
Comment (Shane32) [inline src/GraphQL/Instrumentation/FieldMiddlewareBuilderExtensions.cs]:
```suggestion
                throw new ArgumentException($"Field middleware of type '{middleware.FullName}' must implement the {nameof(IFieldMiddleware)} interface", nameof(middleware));
```

File: docs2/site/docs/getting-started/field-middleware.md [hunk 6]
```diff
@@ -18,10 +18,10 @@ await schema.ExecuteAsync(_ =>
 });
 ```
 
-You can write a class that has a `Resolve` method:
+And write a class that implements `IFieldMiddleware`:
```
Comment (sungam3r) [inline docs2/site/docs/getting-started/field-middleware.md]: Logically it should go first and than section about _You register Field Middleware in `ExecutionOptions`._

Comment (Shane32) [reply docs2/site/docs/getting-started/field-middleware.md]: I agree

Comment (Shane32) [reply docs2/site/docs/getting-started/field-middleware.md]: Take a look now

File: docs2/site/docs/getting-started/field-middleware.md [hunk 7]
```diff
@@ -41,10 +31,20 @@ public class InstrumentFieldsMiddleware
 }
 ```
 
-Or you can register a middleware delegate directly:
+Then register your Field Middleware in `ExecutionOptions`.
+
+```csharp
+await schema.ExecuteAsync(_ =>
```
Comment (sungam3r) [inline docs2/site/docs/getting-started/field-middleware.md]: I propose to change `_` to `options`.

**Comment by sungam3r** [comment]
Everything suits me. Let's wait for the opinions of the others for a while?

**Comment by sungam3r** [comment]
No feedback (at least negative) so merging.

**Comment by Shane32** [comment]
I've already got the middleware covered in the upgrade guide, btw

**Comment by sungam3r** [comment]
I wonder if you use fuget.org, the link to which I made in the readme?

**Comment by Shane32** [comment]
I have not seen that website, but I'll take a look.

**Comment by sungam3r** [comment]
https://www.fuget.org/packages/GraphQL/3.0.0-preview-1352/lib/netstandard2.0/diff/2.4.0/

**Pull request merged**